File: leetcode/go/386.lexicographical-numbers.go
func lexicalOrder(n int) []int {
    ans := make([]int, 0, n+1)
    for i:=1; i < 10 && i <= n; i++ {
        genNums(i, n, &ans)
    }
    return ans
}

func genNums(i int, n int, ans *[]int) {
    (*ans) = append(*ans, i)
    for j :=i * 10; j < i * 10 + 10; j++ {
        if j > n {
            break
        }
        genNums(j, n, ans)
    }
}

File: leetcode/rust-leetcode/src/386.lexicographical-numbers.rs
impl Solution {
    pub fn lexical_order(n: i32) -> Vec<i32> {
        fn addNums(x: i32, n: i32, ans: &mut Vec<i32>) {
            ans.push(x);
            let y = x * 10;
            for i in y..y+10 {
                if i > n {
                    break;
                }
                addNums(i, n, ans);
            }
        }
        let mut ans = Vec::with_capacity(n as usize + 1);
        for i in 1..10 {
            if i > n {
                break;
            }
            addNums(i, n, &mut ans);
        }
        ans
    }
}

